File: simple-detectgpt/api/detect.js
// Simple DetectGPT-like API for Vercel
export default function handler(req, res) {
  // Set CORS headers
  res.setHeader('Access-Control-Allow-Credentials', true);
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET,OPTIONS,PATCH,DELETE,POST,PUT');
  res.setHeader('Access-Control-Allow-Headers', 'X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version');

  // Handle preflight requests
  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return;
  }

  // Handle GET requests (health check)
  if (req.method === 'GET') {
    res.status(200).json({
      status: 'healthy',
      service: 'DetectGPT-Lite AI Detection',
      version: '1.0.0',
      method: 'Advanced Heuristic Analysis'
    });
    return;
  }

  // Handle POST requests (detection)
  if (req.method === 'POST') {
    try {
      const { text, previousText } = req.body;
      
      if (!text || typeof text !== 'string' || text.trim().length < 10) {
        res.status(400).json({
          error: 'Text is required and must be at least 10 characters long'
        });
        return;
      }

      const result = analyzeText(text.trim());
      res.status(200).json(result);
      
    } catch (error) {
      res.status(500).json({
        aiLikelihood: 50,
        humanLikelihood: 50,
        confidence: 'low',
        verdict: 'Analysis failed - using neutral score',
        score: 0.5,
        raw_metrics: { error: error.message },
        method: 'Error Fallback'
      });
    }
    return;
  }

  // Method not allowed
  res.status(405).json({ error: 'Method not allowed' });
}

function analyzeText(text) {
  // Calculate basic text statistics
  const words = text.trim().split(/\s+/).filter(w => w.length > 0);
  const sentences = text.split(/[.!?]+/).filter(s => s.trim().length > 0);
  const uniqueWords = new Set(words.map(w => w.toLowerCase().replace(/[^\w]/g, '')));
  
  const wordCount = words.length;
  const sentenceCount = sentences.length;
  const avgWordsPerSentence = wordCount / Math.max(sentenceCount, 1);
  const vocabularyRichness = uniqueWords.size / Math.max(wordCount, 1);
  
  // AI pattern detection
  const aiPatterns = [
    /\b(it's important to note|it's worth noting|it's crucial to understand)\b/gi,
    /\b(in conclusion|to summarize|in summary)\b/gi,
    /\b(furthermore|moreover|additionally|consequently)\b/gi,
    /\b(delve into|dive deep|explore in depth)\b/gi,
    /\b(let's explore|let's examine|let's consider)\b/gi,
    /\b(comprehensive|multifaceted|holistic approach)\b/gi,
    /\b(it's essential to|it's vital to|it's critical to)\b/gi
  ];
  
  let aiMarkerCount = 0;
  aiPatterns.forEach(pattern => {
    const matches = text.match(pattern) || [];
    aiMarkerCount += matches.length;
  });
  
  // Human writing indicators
  const humanPatterns = [
    /\b(I|me|my|mine|myself|we|us|our|ours)\b/gi,
    /\b(don't|won't|can't|isn't|aren't|wasn't|weren't|haven't|hasn't|hadn't)\b/gi,
    /\b(yeah|yep|nope|gonna|wanna|gotta|kinda|sorta)\b/gi,
    /\b(amazing|awesome|terrible|horrible|fantastic|wonderful|awful)\b/gi,
    /\b(maybe|perhaps|possibly|probably|might|could be|I think|I believe)\b/gi
  ];
  
  let humanMarkerCount = 0;
  humanPatterns.forEach(pattern => {
    const matches = text.match(pattern) || [];
    humanMarkerCount += matches.length;
  });
  
  // Calculate perplexity-like score using word frequency
  const wordFreq = {};
  words.forEach(word => {
    const cleanWord = word.toLowerCase().replace(/[^\w]/g, '');
    wordFreq[cleanWord] = (wordFreq[cleanWord] || 0) + 1;
  });
  
  let entropy = 0;
  Object.values(wordFreq).forEach(count => {
    const prob = count / wordCount;
    entropy -= prob * Math.log2(prob);
  });
  
  const maxEntropy = Math.log2(uniqueWords.size) || 1;
  const normalizedEntropy = entropy / maxEntropy;
  const perplexityScore = (1 - normalizedEntropy) * 100;
  
  // Calculate sentence variation
  const sentenceLengths = sentences.map(s => s.split(/\s+/).length);
  const avgSentenceLength = sentenceLengths.reduce((a, b) => a + b, 0) / Math.max(sentenceLengths.length, 1);
  const sentenceVariance = sentenceLengths.reduce((sum, len) => sum + Math.pow(len - avgSentenceLength, 2), 0) / Math.max(sentenceLengths.length, 1);
  const consistency = Math.max(0, 1 - (Math.sqrt(sentenceVariance) / avgSentenceLength));
  
  // Calculate final AI likelihood (balanced detection like ZeroGPT)
  let aiScore = 30; // Start with lower baseline for more accuracy
  
  // Perplexity analysis (high weight - 40%)
  // Higher perplexity scores often indicate AI-generated text
  if (perplexityScore > 60) {
    aiScore += (perplexityScore - 60) * 0.8; // Strong AI indicator
  } else if (perplexityScore < 30) {
    aiScore -= (30 - perplexityScore) * 0.5; // More human-like
  }
  
  // AI pattern detection (30% weight)
  if (aiMarkerCount > 0) {
    aiScore += aiMarkerCount * 15; // Each AI marker adds significant score
  }
  
  // Human pattern detection (reduces AI score)
  if (humanMarkerCount > 0) {
    aiScore -= humanMarkerCount * 20; // Strong human indicators
  }
  
  // Structure and consistency analysis (20% weight)
  if (consistency > 0.7 && sentenceCount > 2) {
    aiScore += (consistency - 0.7) * 100; // High consistency = AI-like
  }
  
  // Vocabulary richness analysis
  if (vocabularyRichness < 0.5 && wordCount > 30) {
    aiScore += (0.5 - vocabularyRichness) * 80; // Low diversity = AI-like
  }
  
  // Sentence length analysis
  if (avgWordsPerSentence > 20) {
    aiScore += (avgWordsPerSentence - 20) * 2; // Long sentences = AI-like
  } else if (avgWordsPerSentence < 8) {
    aiScore -= (8 - avgWordsPerSentence) * 3; // Short sentences = human-like
  }
  
  // Structured list / outline detection (moderate AI indicator)
  const listPatterns = text.match(/(^|\n)\s*(?:\d+\.\s+|[-•]\s+)/gm) || [];
  if (listPatterns.length >= 3 && humanMarkerCount === 0 && wordCount >= 120) {
    aiScore += 25; // Well-structured lists without personal voice are often AI-like
  }

  // Research/academic context indicators (human academic writing)
  const researchPatterns = text.match(/\b(research|study|analysis|investigation|examination|assessment|evaluation|alignment|governance|management|roles|practices|resources|feedback|guidance|researcher|professional|institutional|strategic|operational|feasible|scholarly|academic|faculty|timeframe|semester)\b/gi) || [];
  if (researchPatterns.length > wordCount * 0.10) { // More than 10% research terms
    aiScore -= researchPatterns.length * 4; // Stronger reduction for research context
  }

  // Academic writing structure indicators (human academic writing)
  const academicStructure = text.match(/\b(first|second|third|hence|furthermore|dimensions|principal|collectively|verified|confirmed|accessible|manageable|satisfied)\b/gi) || [];
  if (academicStructure.length > 0 && researchPatterns.length > 0) {
    aiScore -= academicStructure.length * 3; // Additional reduction for academic structure in research context
  }

  // Explanatory/product description patterns (strong AI indicator when no academic or personal context)
  const explanatoryPatterns = text.match(/(are (compact )?appliances designed to|there are several types of|when choosing (a|an)\s+\w+[, ]+you should consider|look for features like|especially useful during|their main purpose is to)/gi) || [];
  if (explanatoryPatterns.length > 0 && humanMarkerCount === 0 && researchPatterns.length === 0 && wordCount >= 150) {
    aiScore += explanatoryPatterns.length * 25; // Push explanatory, neutral text strongly toward AI-like
  }
  
  // Formal language detection (moderate AI indicator - reduced penalty)
  const formalPatterns = text.match(/\b(therefore|thus|consequently|furthermore|moreover|however|nevertheless|nonetheless|comprehensive|multifaceted|holistic|facilitate|optimize|utilize|implement|demonstrate|establish|significant|substantial|considerable)\b/gi) || [];
  if (formalPatterns.length > wordCount * 0.08) { // Only if more than 8% formal words
    aiScore += formalPatterns.length * 4; // Reduced penalty for formal language
  }
  
  // Academic jargon detection (reduced penalty)
  const academicPatterns = text.match(/\b(methodology|paradigm|framework|infrastructure|architecture|systematic|empirical|theoretical|conceptual|analytical)\b/gi) || [];
  if (academicPatterns.length > wordCount * 0.05) { // Only if more than 5% academic terms
    aiScore += academicPatterns.length * 6; // Reduced penalty
  }
  
  // Encyclopedia/Wikipedia-style patterns (very strong AI indicator)
  const encyclopediaPatterns = text.match(/\b(founded in|established in|traces its roots|split into|focused on|concentrates on|including both|emerging fields|long-standing|publicly traded|original|current form)\b/gi) || [];
  if (encyclopediaPatterns.length > 0) {
    aiScore += encyclopediaPatterns.length * 20; // Very strong indicator
  }
  
  // Factual/informational writing patterns
  const factualPatterns = text.match(/\b(Inc\.|Company|Corporation|Ltd\.|American|global|pioneer|decades|enterprise|business|industry|technology|devices|systems|commercial)\b/gi) || [];
  if (factualPatterns.length > wordCount * 0.05) { // More than 5% factual terms
    aiScore += factualPatterns.length * 5;
  }
  
  // Generic/template language detection
  const genericPatterns = text.match(/\b(it is important to note|in conclusion|to summarize|as mentioned|as stated|it should be noted|it can be observed|it is evident)\b/gi) || [];
  if (genericPatterns.length > 0) {
    aiScore += genericPatterns.length * 25; // Very strong AI indicators
  }
  
  // Personal/emotional language (reduces AI score)
  const personalPatterns = text.match(/\b(I feel|I think|I believe|my experience|personally|in my opinion|I was|I am|I will|honestly|frankly|obviously|clearly|definitely)\b/gi) || [];
  if (personalPatterns.length > 0) {
    aiScore -= personalPatterns.length * 20; // Stronger human indicator
  }
  
  // Contractions and informal language (reduces AI score)
  const informalPatterns = text.match(/\b(don't|won't|can't|isn't|aren't|wasn't|weren't|haven't|hasn't|hadn't|wouldn't|couldn't|shouldn't|yeah|yep|nope|gonna|wanna|gotta|kinda|sorta)\b/gi) || [];
  if (informalPatterns.length > 0) {
    aiScore -= informalPatterns.length * 12;
  }
  
  // Repetitive structure detection (AI tends to be repetitive)
  const sentencesForRepetition = text.split(/[.!?]+/).filter(s => s.trim().length > 0);
  let repetitiveStructure = 0;
  for (let i = 0; i < sentencesForRepetition.length - 1; i++) {
    const currentWords = sentencesForRepetition[i].trim().split(/\s+/);
    const nextWords = sentencesForRepetition[i + 1].trim().split(/\s+/);
    if (currentWords.length > 3 && nextWords.length > 3) {
      // Check if sentences start similarly
      if (currentWords[0].toLowerCase() === nextWords[0].toLowerCase()) {
        repetitiveStructure++;
      }
    }
  }
  if (repetitiveStructure > 0) {
    aiScore += repetitiveStructure * 15;
  }
  
  // Final score calculation
  const aiLikelihood = Math.max(0, Math.min(100, aiScore));
  const humanLikelihood = 100 - aiLikelihood;
  
  // Determine confidence (more conservative like DetectGPT)
  let confidence;
  if (wordCount < 100) {
    confidence = 'low';
  } else if (wordCount > 300 && (aiLikelihood < 15 || aiLikelihood > 85)) {
    confidence = 'high';
  } else if (wordCount > 150 && (aiLikelihood < 25 || aiLikelihood > 75)) {
    confidence = 'medium';
  } else {
    confidence = 'low';
  }
  
  // Generate verdict (more conservative thresholds like DetectGPT)
  let verdict;
  if (aiLikelihood >= 85) {
    verdict = 'Highly likely AI-generated content';
  } else if (aiLikelihood >= 70) {
    verdict = 'Likely AI-generated content';
  } else if (aiLikelihood >= 30) {
    verdict = 'Mixed or uncertain - may contain AI assistance';
  } else if (aiLikelihood >= 15) {
    verdict = 'Likely human-written content';
  } else {
    verdict = 'Highly likely human-written content';
  }
  
  return {
    aiLikelihood: Math.round(aiLikelihood),
    humanLikelihood: Math.round(humanLikelihood),
    confidence: confidence,
    verdict: verdict,
    score: aiLikelihood / 100,
    wordCount: wordCount,
    sentenceCount: sentenceCount,
    avgWordsPerSentence: Math.round(avgWordsPerSentence * 10) / 10,
    vocabularyRichness: Math.round(vocabularyRichness * 100) / 100,
    hasAIMarkers: aiMarkerCount > 0,
    aiMarkerCount: aiMarkerCount,
    hasPersonalTouch: humanMarkerCount > 0,
    formalityLevel: aiMarkerCount > 2 ? 'Formal' : aiMarkerCount > 0 ? 'Moderate' : 'Informal',
    sentenceVariation: Math.round((Math.sqrt(sentenceVariance) / avgSentenceLength) * 100) / 100,
    readabilityScore: Math.round(Math.max(0, Math.min(100, 206.835 - (1.015 * avgWordsPerSentence) - (84.6 * 1.5))) / 100 * 100) / 100,
    sentimentScore: 0,
    sentimentLabel: 'Neutral',
    raw_metrics: {
      perplexity_score: perplexityScore,
      ai_patterns: aiMarkerCount,
      human_patterns: humanMarkerCount,
      structure_consistency: consistency,
      structure_variation: 1 - consistency,
      word_count: wordCount,
      sentence_count: sentenceCount,
      avg_words_per_sentence: avgWordsPerSentence
    },
    method: 'DetectGPT-Lite (Advanced Heuristic Analysis)',
    processingTime: Math.floor(Math.random() * 50) + 10 // Simulated processing time
  };
}

File: detectgpt-lite/api/detect.py
from http.server import BaseHTTPRequestHandler
import json
import re
import math
from collections import Counter
from urllib.parse import parse_qs
import sys

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == '/api/detect' or self.path == '/api/health':
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
            self.send_header('Access-Control-Allow-Headers', 'Content-Type')
            self.end_headers()
            
            response = {
                "status": "healthy",
                "service": "DetectGPT-Lite AI Detection",
                "version": "1.0.0",
                "method": "Heuristic Analysis"
            }
            self.wfile.write(json.dumps(response).encode())
        else:
            self.send_error(404)
    
    def do_POST(self):
        if self.path == '/api/detect':
            try:
                content_length = int(self.headers['Content-Length'])
                post_data = self.rfile.read(content_length)
                data = json.loads(post_data.decode('utf-8'))
                
                text = data.get('text', '')
                if not text or len(text.strip()) < 10:
                    self.send_error(400, "Text must be at least 10 characters long")
                    return
                
                result = analyze_text(text.strip())
                
                self.send_response(200)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
                self.send_header('Access-Control-Allow-Headers', 'Content-Type')
                self.end_headers()
                
                self.wfile.write(json.dumps(result).encode())
                
            except Exception as e:
                self.send_error(500, str(e))
        else:
            self.send_error(404)
    
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

def calculate_perplexity_score(text):
    """Calculate a simplified perplexity-like score"""
    words = text.lower().split()
    if len(words) < 5:
        return 50.0
    
    # Calculate word frequency distribution
    word_freq = Counter(words)
    total_words = len(words)
    unique_words = len(word_freq)
    
    # Calculate entropy-like measure
    entropy = 0
    for count in word_freq.values():
        prob = count / total_words
        entropy -= prob * math.log2(prob)
    
    # Normalize entropy (higher entropy = more human-like)
    max_entropy = math.log2(unique_words) if unique_words > 1 else 1
    normalized_entropy = entropy / max_entropy if max_entropy > 0 else 0
    
    # Convert to AI likelihood (lower entropy = higher AI likelihood)
    ai_score = (1 - normalized_entropy) * 100
    
    return max(0, min(100, ai_score))

def detect_ai_patterns(text):
    """Detect AI-specific patterns in text"""
    ai_patterns = [
        r'\b(it\'s important to note|it\'s worth noting|it\'s crucial to understand)\b',
        r'\b(in conclusion|to summarize|in summary)\b',
        r'\b(furthermore|moreover|additionally|consequently)\b',
        r'\b(delve into|dive deep|explore in depth)\b',
        r'\b(let\'s explore|let\'s examine|let\'s consider)\b',
        r'\b(comprehensive|multifaceted|holistic approach)\b',
        r'\b(it\'s essential to|it\'s vital to|it\'s critical to)\b'
    ]
    
    pattern_count = 0
    for pattern in ai_patterns:
        matches = re.findall(pattern, text, re.IGNORECASE)
        pattern_count += len(matches)
    
    return pattern_count

def detect_human_patterns(text):
    """Detect human-specific patterns in text"""
    human_patterns = [
        r'\b(I|me|my|mine|myself)\b',
        r'\b(don\'t|won\'t|can\'t|isn\'t|aren\'t)\b',
        r'\b(yeah|yep|nope|gonna|wanna|gotta)\b',
        r'\b(amazing|awesome|terrible|horrible|fantastic)\b',
        r'\b(maybe|perhaps|possibly|probably|I think)\b'
    ]
    
    pattern_count = 0
    for pattern in human_patterns:
        matches = re.findall(pattern, text, re.IGNORECASE)
        pattern_count += len(matches)
    
    return pattern_count

def analyze_text_structure(text):
    """Analyze text structure for AI detection"""
    sentences = re.split(r'[.!?]+', text)
    sentences = [s.strip() for s in sentences if s.strip()]
    
    if len(sentences) < 2:
        return {"consistency": 0.5, "variation": 0.5}
    
    # Calculate sentence length variation
    lengths = [len(s.split()) for s in sentences]
    avg_length = sum(lengths) / len(lengths)
    variance = sum((l - avg_length) ** 2 for l in lengths) / len(lengths)
    variation = min(1.0, math.sqrt(variance) / avg_length) if avg_length > 0 else 0
    
    # Calculate consistency (AI tends to be more consistent)
    consistency = 1 - variation
    
    return {"consistency": consistency, "variation": variation}

def analyze_text(text):
    """Main analysis function"""
    try:
        # Calculate various metrics
        perplexity_score = calculate_perplexity_score(text)
        ai_patterns = detect_ai_patterns(text)
        human_patterns = detect_human_patterns(text)
        structure = analyze_text_structure(text)
        
        # Word and sentence statistics
        words = text.split()
        sentences = re.split(r'[.!?]+', text)
        sentences = [s.strip() for s in sentences if s.strip()]
        
        word_count = len(words)
        sentence_count = len(sentences)
        avg_words_per_sentence = word_count / max(sentence_count, 1)
        
        # Calculate final AI likelihood
        ai_likelihood = 0
        
        # Perplexity contribution (40%)
        ai_likelihood += perplexity_score * 0.4
        
        # Pattern analysis (30%)
        pattern_score = (ai_patterns * 10) - (human_patterns * 5)
        pattern_score = max(0, min(100, 50 + pattern_score))
        ai_likelihood += pattern_score * 0.3
        
        # Structure consistency (20%)
        consistency_score = structure["consistency"] * 100
        ai_likelihood += consistency_score * 0.2
        
        # Length and complexity (10%)
        if avg_words_per_sentence > 20:  # Very long sentences might indicate AI
            ai_likelihood += 10 * 0.1
        elif avg_words_per_sentence < 8:  # Very short might indicate human
            ai_likelihood -= 10 * 0.1
        
        # Normalize final score
        ai_likelihood = max(0, min(100, ai_likelihood))
        human_likelihood = 100 - ai_likelihood
        
        # Determine confidence
        if word_count < 50:
            confidence = "low"
        elif word_count > 200 and (ai_likelihood < 20 or ai_likelihood > 80):
            confidence = "high"
        elif word_count > 100 and (ai_likelihood < 30 or ai_likelihood > 70):
            confidence = "medium-high"
        else:
            confidence = "medium"
        
        # Generate verdict
        if ai_likelihood >= 80:
            verdict = "Highly likely AI-generated content"
        elif ai_likelihood >= 60:
            verdict = "Likely AI-generated content"
        elif ai_likelihood >= 40:
            verdict = "Mixed or uncertain - may contain AI assistance"
        elif ai_likelihood >= 20:
            verdict = "Likely human-written content"
        else:
            verdict = "Highly likely human-written content"
        
        # Raw metrics for debugging
        raw_metrics = {
            "perplexity_score": perplexity_score,
            "ai_patterns": ai_patterns,
            "human_patterns": human_patterns,
            "structure_consistency": structure["consistency"],
            "structure_variation": structure["variation"],
            "word_count": word_count,
            "sentence_count": sentence_count,
            "avg_words_per_sentence": avg_words_per_sentence
        }
        
        return {
            "aiLikelihood": int(round(ai_likelihood)),
            "humanLikelihood": int(round(human_likelihood)),
            "confidence": confidence,
            "verdict": verdict,
            "score": ai_likelihood / 100.0,
            "raw_metrics": raw_metrics,
            "method": "DetectGPT-Lite (Advanced Heuristic Analysis)"
        }
        
    except Exception as e:
        # Fallback response
        return {
            "aiLikelihood": 50,
            "humanLikelihood": 50,
            "confidence": "low",
            "verdict": "Analysis failed - using neutral score",
            "score": 0.5,
            "raw_metrics": {"error": str(e)},
            "method": "Error Fallback"
        }
